### ISSUE TITLE
[rcm-03] Criação do Modelo

### ISSUE THREAD

**Issue by GuiRocDuarte** [jobsrobson/ProjetoDeepLearning#3]
- [x] Dimensionar as imagens
- [ ] Normalização dos valores dos pixels (0 a 1)
- [ ] Escolher o modelo (criar ou usar um pronto?)
- [ ] Treinar o modelo

**Comment by GuiRocDuarte** [comment]
- [ ] Dimensionar as imagens

**Comment by jobsrobson** [comment]
Modelo utilizado pelos autores do dataset: [InceptionV4](https://github.com/kentsommer/keras-inceptionV4/tree/ef1db6f09b6511779c05fab47d374741bc89b5ee)

**Comment by jobsrobson** [comment]
O InceptionV4 é uma arquitetura de rede neural convolucional avançada e de alto desempenho, desenvolvida pelo Google. Ele é uma evolução do InceptionV3 e foi apresentado junto com as arquiteturas Inception-ResNet.

Pontos Positivos do InceptionV4:

Desempenho de Ponta: O InceptionV4, assim como as variações Inception-ResNet, alcançou resultados muito competitivos em benchmarks como o ImageNet. Ele é mais complexo e, em geral, mais poderoso que o InceptionV3.
Arquitetura Otimizada: Ele apresenta uma arquitetura mais "uniforme" e "limpa" em comparação com suas versões anteriores (como o InceptionV3, que acumulou algumas "heranças" de versões mais antigas). Isso pode levar a um treinamento mais estável e, potencialmente, a melhores resultados.
Conexões Residuais (Inception-ResNet): Embora o InceptionV4 "puro" não tenha conexões residuais, ele foi desenvolvido em paralelo com o Inception-ResNet, que incorpora os conceitos de skip connections (pulos de camadas) das redes ResNet. Essas conexões residuais são cruciais para treinar redes muito profundas sem degradação do desempenho, ajudando a mitigar o problema do vanishing gradient.
Pontos de Atenção/Desafios:

Implementação no Keras: O InceptionV4 não faz parte diretamente do módulo tf.keras.applications oficial (que inclui InceptionV3, ResNet, Xception, EfficientNet, etc.). Isso significa que você não pode simplesmente importar keras.applications.InceptionV4. Geralmente, as implementações de InceptionV4 no Keras são de repositórios de terceiros (como o que você provavelmente está se referindo, kentsommer/keras-inceptionV4 no GitHub).
Implicação: Você precisará baixar e usar o código dessa implementação externa, e carregar os pesos pré-treinados fornecidos por ela. Isso exige um pouco mais de cuidado na configuração do que usar um modelo nativo do Keras.
Complexidade e Recursos: Por ser mais complexo, o InceptionV4 pode exigir mais recursos computacionais (GPU/CPU, RAM) e levar mais tempo para treinar em comparação com modelos mais leves. No entanto, sua base de dados de 299x299 é a resolução ideal para ele.

Comparação com InceptionV3: Embora o InceptionV4 seja teoricamente superior, em alguns casos (especialmente com datasets menores ou para tarefas específicas), a diferença de desempenho para o InceptionV3 pode não ser tão drástica a ponto de justificar a complexidade adicional de uma implementação de terceiros. Para alguns datasets, o InceptionV3 ou até mesmo o Inception-ResNetV2 podem ter um desempenho similar ou até ligeiramente melhor, dependendo dos detalhes.

Como Implementar o keras-inceptionV4:

Você provavelmente precisará clonar ou baixar o código da implementação de InceptionV4 de um repositório como o kentsommer/keras-inceptionV4 no GitHub.
As instruções de uso e como carregar os pesos pré-treinados (weights='imagenet') estarão no README desse repositório. O processo é similar ao que te mostrei, mas em vez de importar InceptionV3 do tf.keras.applications, você importaria a função de criação do modelo daquele arquivo específico (e.g., from keras_inceptionv4 import inception_v4).
Preste atenção aos requisitos de pré-processamento de entrada do InceptionV4 daquela implementação específica (normalização dos pixels, por exemplo, pode ser diferente do rescale=1./255).
Mantenha o Fine-Tuning: A estratégia de transfer learning (congelar camadas iniciais, treinar as novas e depois fazer fine-tuning) continua sendo a melhor abordagem.

**Comment by jobsrobson** [comment]
import tensorflow as tf
from tensorflow.keras.preprocessing.image import ImageDataGenerator
from tensorflow.keras.layers import Dense, GlobalAveragePooling2D
from tensorflow.keras.models import Model
from tensorflow.keras.optimizers import Adam
import os
import numpy as np # Necessário para a função preprocess_input
import matplotlib.pyplot as plt

# Importa a função create_model do arquivo inception_v4.py
# Certifique-se de que 'inception_v4.py' está no mesmo diretório do seu script
from inception_v4 import create_model, preprocess_input

# --- Configurações Iniciais ---
DATA_DIR = 'caminho/para/sua/pasta_com_classes_de_resistores' # MUDE ISSO PARA O SEU DIRETÓRIO!
IMAGE_SIZE = (299, 299)
BATCH_SIZE = 32
NUM_CLASSES = 37 # Confirme o número exato de classes após a unificação!

# --- 1. Preparação dos Geradores de Dados ---
# ATENÇÃO: Removemos 'rescale=1./255' porque o InceptionV4 tem seu próprio pré-processamento.
# Em vez disso, usaremos a função 'preprocess_input' fornecida pelo modelo.

# Gerador para treino (com augmentation e split para validação)
train_datagen = ImageDataGenerator(
    rotation_range=15, # Ajuste conforme necessário
    width_shift_range=0.1,
    height_shift_range=0.1,
    brightness_range=[0.7, 1.3],
    shear_range=0.1,
    zoom_range=0.1,
    horizontal_flip=True,
    validation_split=0.20, # 20% para validação
    preprocessing_function=preprocess_input # APLICA A FUNÇÃO DE PRÉ-PROCESSAMENTO DO INCEPTIONV4
)

# Gerador para validação (sem augmentation, apenas pré-processamento)
# Note que estamos usando o mesmo train_datagen para gerar a validação via subset
validation_generator = train_datagen.flow_from_directory(
    DATA_DIR,
    target_size=IMAGE_SIZE,
    color_mode='rgb',
    batch_size=BATCH_SIZE,
    class_mode='categorical',
    subset='validation', # Para o conjunto de validação
    seed=42
)

# Gerador para treino
train_generator = train_datagen.flow_from_directory(
    DATA_DIR,
    target_size=IMAGE_SIZE,
    color_mode='rgb',
    batch_size=BATCH_SIZE,
    class_mode='categorical',
    subset='training', # Para o conjunto de treino
    seed=42
)

# Confirma o número de classes que o gerador detectou
print(f"Número de classes detectadas no treino: {train_generator.num_classes}")
print(f"Classes detectadas: {train_generator.class_indices}")
if train_generator.num_classes != NUM_CLASSES:
    print(f"AVISO: O gerador de dados detectou {train_generator.num_classes} classes, mas NUM_CLASSES está configurado como {NUM_CLASSES}.")
    print("Por favor, verifique e corrija NUM_CLASSES se necessário.")
    NUM_CLASSES = train_generator.num_classes # Ajusta NUM_CLASSES se houver divergência

# --- 2. Construção do Modelo com InceptionV4 e Transfer Learning ---

# Cria o modelo InceptionV4. 'include_top=False' remove as camadas finais de classificação.
# 'weights='imagenet'' carrega os pesos pré-treinados do ImageNet.
print("Carregando modelo InceptionV4 com pesos do ImageNet...")
try:
    base_model = create_model(
        weights='imagenet',       # Carrega os pesos pré-treinados do ImageNet
        include_top=False         # Não inclui a camada de classificação final do ImageNet
                                  # Não é necessário passar input_shape aqui, a função create_model já define
    )
    print("Modelo InceptionV4 carregado com sucesso!")
except Exception as e:
    print(f"Erro ao carregar InceptionV4: {e}")
    print("Verifique se o arquivo 'inception_v4.py' está no local correto e se há conexão com a internet para baixar os pesos.")
    print("Se o problema persistir, considere tentar com InceptionV3 ou Xception para depuração.")
    exit()

# Congela as camadas do modelo base para que não sejam treinadas na primeira fase
for layer in base_model.layers:
    layer.trainable = False

# Adiciona novas camadas de classificação para o seu problema
x = base_model.output
x = GlobalAveragePooling2D()(x) # Converte os feature maps para um vetor
x = Dense(512, activation='relu')(x) # Camada densa com ReLU para não-linearidade
predictions = Dense(NUM_CLASSES, activation='softmax')(x) # Camada de saída com softmax para classificação multi-classe

# Cria o modelo final
model = Model(inputs=base_model.input, outputs=predictions)

# --- 3. Compilação e Treinamento (Fase 1: Apenas novas camadas) ---
model.compile(optimizer=Adam(learning_rate=0.001),
              loss='categorical_crossentropy',
              metrics=['accuracy'])

print("\n--- Iniciando o treinamento da Fase 1 (apenas novas camadas) ---")
# Use callbacks para EarlyStopping e ModelCheckpoint
callbacks = [
    tf.keras.callbacks.EarlyStopping(monitor='val_loss', patience=10, restore_best_weights=True),
    tf.keras.callbacks.ModelCheckpoint('best_model_phase1.h5', save_best_only=True, monitor='val_accuracy', mode='max')
]

history_phase1 = model.fit(
    train_generator,
    epochs=20, # Comece com um número razoável de épocas
    validation_data=validation_generator,
    callbacks=callbacks
)

# --- 4. Fine-tuning (Fase 2: Descongelar e treinar parte do modelo base) ---
print("\n--- Iniciando o Fine-tuning (Fase 2) ---")

# Descongele as camadas do modelo base para fine-tuning
# O InceptionV4 é bem profundo. Você pode descongelar todas, ou um subconjunto.
# Para este dataset, vamos tentar descongelar a partir de uma camada mais profunda.
# Inspecione `base_model.summary()` para ver os nomes das camadas e decidir onde descongelar.
# Por exemplo, se quiser descongelar tudo:
for layer in base_model.layers:
    layer.trainable = True

# Ou um exemplo mais seletivo:
# last_conv_layer_name = 'conv_7b_2_bn' # Exemplo de nome de camada InceptionV4 para descongelar a partir dela
# set_trainable = False
# for layer in base_model.layers:
#     if layer.name == last_conv_layer_name:
#         set_trainable = True
#     if set_trainable:
#         layer.trainable = True
#     else:
#         layer.trainable = False

# Recompile o modelo com uma taxa de aprendizado muito menor para fine-tuning
model.compile(optimizer=Adam(learning_rate=0.00001), # Taxa de aprendizado bem menor!
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Use callbacks para EarlyStopping e ModelCheckpoint para a Fase 2 também
callbacks_fine_tune = [
    tf.keras.callbacks.EarlyStopping(monitor='val_loss', patience=15, restore_best_weights=True), # Paciência maior
    tf.keras.callbacks.ModelCheckpoint('best_model_fine_tuned.h5', save_best_only=True, monitor='val_accuracy', mode='max')
]

history_phase2 = model.fit(
    train_generator,
    epochs=55, # Total de épocas, digamos 20 da fase 1 + 35 para fase 2
    initial_epoch=history_phase1.epoch[-1] + 1, # Continua a partir da última época da fase 1
    validation_data=validation_generator,
    callbacks=callbacks_fine_tune
)

# --- 5. Salvar o Modelo Final ---
model.save('modelo_resistores_inceptionv4_final.h5')
print("\nModelo treinado e salvo como 'modelo_resistores_inceptionv4_final.h5'")

# --- 6. Visualizar o Histórico de Treinamento ---
# Combine os históricos
acc = history_phase1.history['accuracy'] + history_phase2.history['accuracy']
val_acc = history_phase1.history['val_accuracy'] + history_phase2.history['val_accuracy']
loss = history_phase1.history['loss'] + history_phase2.history['loss']
val_loss = history_phase1.history['val_loss'] + history_phase2.history['val_loss']

plt.figure(figsize=(12, 4))
plt.subplot(1, 2, 1)
plt.plot(acc, label='Acurácia de Treino')
plt.plot(val_acc, label='Acurácia de Validação')
plt.title('Acurácia do Modelo')
plt.xlabel('Época')
plt.ylabel('Acurácia')
plt.legend()

plt.subplot(1, 2, 2)
plt.plot(loss, label='Perda de Treino')
plt.plot(val_loss, label='Perda de Validação')
plt.title('Perda do Modelo')
plt.xlabel('Época')
plt.ylabel('Perda')
plt.legend()
plt.show()

print("\n--- Próximo Passo: Avaliação Detalhada e Criação do Conjunto de Teste Separado ---")
